github-actions/slash-commands/main.ts

```diff
@@ -0,0 +1,90 @@
+import * as core from '@actions/core';
+import {context} from '@actions/github';
+import {rerunCircleCi} from './commands/rerun-circleci';
+import {rebase} from './commands/rebase';
+
+/** The marker used in comments to note a command is being made in the comments. */
+const commandMarker = '/ng-bot';
+/** The full command provided in the comment, split on spaces from the comment. */
+let parsedCommand: [string, ...string[]];
+
+/** Get the parsed command from the comment. */
+function getParsedCommand() {
+  if (parsedCommand) {
+    return parsedCommand;
+  }
+  const comment = context.payload.comment!.body || '';
+  const command = (comment.split('\n')[0] || '')
+    .split(' ')
+    .filter((word: string) => !!word)
+    .map((word: string) => word.trim());
+
+  if (command[0] !== commandMarker) {
+    throw Error();
+  }
+  parsedCommand = command.slice(1);
+
+  return parsedCommand;
+}
+
+/**
+ * Determine whether the action triggers and triggering metadata means a slash command should
+ * actually run.
+ */
+function shouldPerformAction(): boolean {
+  if (context.payload.sender!.type === 'Bot') {
+    core.info(`Skipping as this action was triggered by a bot: ${context.actor}`);
+    return false;
+  }
+  if (context.payload.issue!.pull_request === undefined) {
+    core.info('Skipping as this action was triggered from an issue not a pull request');
+    return false;
+  }
+  if (context.eventName !== 'issue_comment') {
+    core.info('Skipping as this action was not an issue_comment trigger.');
+    return false;
+  }
+  if (!['created', 'edited'].includes(context.payload.action!)) {
+    core.info('Skipping as this action was not a created or edited issue_comment trigger');
+    return false;
+  }
+  if (!context.payload.comment!.body!.startsWith(commandMarker)) {
+    core.info(`Skipping as the comment does not begin with ${commandMarker}`);
+    return false;
+  }
+  if (getParsedCommand().length === 0) {
+    core.info(`Skipping as only the commandMarket (${commandMarker}) is provided as a command`);
+    return false;
+  }
+  return true;
+}
+
+/** The main function for the slash-command action. */
+async function run(): Promise<void> {
+  if (!shouldPerformAction()) {
```

Comment:
Should we also limit the action to the PR author, and people with write access? That's what I did in my SSR github action: https://github.com/angular/components/pull/23032/files.

Related to that, I was also wondering in terms of security: consider a case where one command will deal with the checked-out PR. This is usually prone to leaking the key. See: https://securitylab.github.com/research/github-actions-preventing-pwn-requests/. Maybe this is already a potential issue for the rebase action right now? (e.g. if I set up some Git hooks; which we might be able to prevent though as a side-note)

There is an action that deals with slash commands in some way already: https://github.com/peter-evans/slash-command-dispatch.

Reply:
Updated to include a permissions check.

For security, as discussed offline, https://github.com/angular/dev-infra/pull/162 should help us to create a good platform to build upon


The reason I didn't want to use the slash command actions already created is that they are managed and configured entirely in the `.github/<workflow>.yml` file which means that it is extremely difficult for us to centrally manage in any way since its done in the downstream repo.